src/coreclr/jit/hwintrinsic.cpp

```diff
@@ -541,22 +541,22 @@ NamedIntrinsic HWIntrinsicInfo::lookupId(Compiler*         comp,
         //
         // When the target hardware does support the instruction set, we can return a
         // constant true. When it doesn't then we want to report the check as dynamically
-        // supported instead. This allows some targets, such as AOT, to emit a check against
-        // a cached CPU query so lightup can still happen (such as for SSE4.1 when the target
-        // hardware is SSE2).
+        // supported instead if the opportunistic support does exist. This allows some targets,
+        // such as AOT, to emit a check against a cached CPU query so lightup can still happen
+        // (such as for SSE4.1 when the target hardware is SSE2).
         //
         // When the compiler doesn't support ISA or when it does but the target hardware does
         // not and we aren't in a scenario with support for a dynamic check, we want to return false.
 
-        if (isIsaSupported)
+        if (isIsaSupported && comp->compSupportsHWIntrinsic(isa))
         {
-            if (comp->compExactlyDependsOn(isa))
+            if (!comp->IsTargetAbi(CORINFO_NATIVEAOT_ABI) || comp->compExactlyDependsOn(isa))
```

Comment:
Could you explain what change is happening here? I don't see a description of what we're changing around NativeAOT behavior in the change description.

Reply:
There was an existing bug here that was surfaced if crossgen/naot targeted `avx` but not `avx2`

For most cases, the ISA initially checked and tracked as part of `isIsaSupported` is the same as what is tracked by the `InstructionSetDesc`

However, for `Vector256` in particular we have the case where the implication is on `Avx` and we will accelerate some APIs when only `Avx` is supported. But, we only want `IsHardwareAccelerated` to report `true` when `Avx2` is also supported.

We were then ending up in a scenario where we'd end up failing to handle `IsHardwareAccelerated` for the recursive case when `avx` was supported but `avx2` was not because `isIsaSupported` (`AVX`) would be `true` and then we'd fail the `compExactlyDependsOn` check for `AVX2` and then return `NI_IsSupported_Dynamic`, which was incorrect since `avx2` was not opportunistic.

This fixes that so we now ensure that we only go down the true/dynamic path if the compiler could support `AVX2` at all.